sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryHelpers.java

```diff
@@ -55,6 +59,77 @@
           + " an earlier stage of the pipeline, this validation can be disabled using"
           + " #withoutValidation.";
 
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryHelpers.class);
+
+  // Given a potential failure and a current job-id, return the next job-id to be used on retry.
+  // Algorithm is as follows (given input of job_id_prefix-N)
+  //   If BigQuery has no status for job_id_prefix-n, we should retry with the same id.
+  //   If job-id-prefix-n is in the PENDING or successful states, no retry is needed.
+  //   Otherwise (job-id-prefix-n completed with errors), try again with job-id-prefix-(n+1)
+  //
+  // We continue to loop through these job ids until we find one that has either succeed, or that
+  // has not been issued yet.
+  static class RetryJobIdResult {
+    public final String jobId;
+    public final boolean shouldRetry;
+
+    public RetryJobIdResult(String jobId, boolean shouldRetry) {
+      this.jobId = jobId;
+      this.shouldRetry = shouldRetry;
+    }
+  }
+
+  static RetryJobIdResult getRetryJobId(
+      String currentJobId, String projectId, String bqLocation, JobService jobService)
+      throws InterruptedException {
+    // Job ids should always be of the form <job_id_prefix>-<retry_count>
```

Comment:
Add a BigQueryJobId class with prefix, retry properties to make sure that this invariant is preserved ?

Reply:
Good idea. Done